[rush] Rush deploy hard fails on missing optionalDependencies

## Please prefix the issue title with the project name i.e. [rush], [api-extractor] etc. ##

**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**

Considering this a bug as optionalDependencies of dependencies of dependencies shouldn't be required.

`ERROR: Cannot find module 'fsevents/' from '/path/to/common/temp/node_modules/.pnpm/registry.npmjs.org/chokidar/2.1.8/node_modules/chokidar'`

**What is the expected behavior?**
optionalDependencies to be included if available, skipped otherwise

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** Rush (using pnpm)
* **Tool Version:** 5.24.1
* **Node Version:** 10.20.1 (LTS)
  * **Is this a LTS version?** yes
  * **Have you tested on a LTS version?** yes
* **OS:** Ubuntu 20.04

Related to #1078

--- a/apps/rush-lib/src/logic/deploy/DeployManager.ts
+++ b/apps/rush-lib/src/logic/deploy/DeployManager.ts
@@ -238,46 +238,54 @@ export class DeployManager {
         // to system modules.  Appending a "/" forces require() to load the NPM package.
         const resolveSuffix: string = dependencyPackageName + resolve.isCore(dependencyPackageName) ? '/' : '';
 
-        const resolvedDependency: string = resolve.sync(dependencyPackageName + resolveSuffix, {
-          basedir: packageJsonRealFolderPath,
-          preserveSymlinks: false,
-          packageFilter: (pkg, dir) => {
-            // point "main" at a file that is guaranteed to exist
-            // This helps resolve packages such as @types/node that have no entry point
-            pkg.main = "./package.json";
-            return pkg;
-          },
-          realpathSync: (filePath) => {
-            // This code fragment is a modification of the documented default implementation from the "fs-extra" docs
-            try {
-              const resolvedPath: string = fs.realpathSync(filePath);
+        try {
+          const resolvedDependency: string = resolve.sync(dependencyPackageName + resolveSuffix, {
+            basedir: packageJsonRealFolderPath,
+            preserveSymlinks: false,
+            packageFilter: (pkg, dir) => {
+              // point "main" at a file that is guaranteed to exist
+              // This helps resolve packages such as @types/node that have no entry point
+              pkg.main = "./package.json";
+              return pkg;
+            },
+            realpathSync: (filePath) => {
+              // This code fragment is a modification of the documented default implementation from the "fs-extra" docs
+              try {
+                const resolvedPath: string = fs.realpathSync(filePath);
 
-              subdemploymentState.symlinkAnalyzer.analyzePath(filePath);
-              return resolvedPath;
-            } catch (realpathErr) {
-              if (realpathErr.code !== "ENOENT") {
-                throw realpathErr;
+                subdemploymentState.symlinkAnalyzer.analyzePath(filePath);
+                return resolvedPath;
+              } catch (realpathErr) {
+                if (realpathErr.code !== "ENOENT") {
+                  throw realpathErr;
+                }
               }
-            }
-            return filePath;
-          },
-        });
+              return filePath;
+            },
+          });
 
-        if (!resolvedDependency) {
-          if (optionalDependencyNames.has(dependencyPackageName)) {
+          if (!resolvedDependency) {
+            if (optionalDependencyNames.has(dependencyPackageName)) {
+              // Ignore missing optional dependency
+              continue;
+            }
+            throw new Error(`Error resolving ${dependencyPackageName} from ${packageJsonRealFolderPath}`);
+          }
+
+          const dependencyPackageFolderPath: string | undefined
+              = this._packageJsonLookup.tryGetPackageFolderFor(resolvedDependency);
+          if (!dependencyPackageFolderPath) {
+            throw new Error(`Error finding package.json folder for ${resolvedDependency}`);
+          }
+
+          this._collectFoldersRecursive(dependencyPackageFolderPath, subdemploymentState);
+        } catch (resolveErr) {
+          if (resolveErr.code === 'MODULE_NOT_FOUND' && optionalDependencyNames.has(dependencyPackageName)) {
             // Ignore missing optional dependency
             continue;
           }
-          throw new Error(`Error resolving ${dependencyPackageName} from ${packageJsonRealFolderPath}`);
+          throw resolveErr;
         }
-
-        const dependencyPackageFolderPath: string | undefined
-          = this._packageJsonLookup.tryGetPackageFolderFor(resolvedDependency);
-        if (!dependencyPackageFolderPath) {
-          throw new Error(`Error finding package.json folder for ${resolvedDependency}`);
-        }
-
-        this._collectFoldersRecursive(dependencyPackageFolderPath, subdemploymentState);
       }
     }
   }

--- a/common/changes/@microsoft/rush/fix-optionalDependencies-required-for-deploy_2020-06-02-13-21.json
+++ b/common/changes/@microsoft/rush/fix-optionalDependencies-required-for-deploy_2020-06-02-13-21.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Expect error when trying to resolve optional dependency during deploy",
+      "type": "none"
+    }
+  ],
+  "packageName": "@microsoft/rush",
+  "email": "jlsjonas@users.noreply.github.com"
+}

